<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell immune repertoire trajectory analysis with dandelionR • dandelionR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell immune repertoire trajectory analysis with dandelionR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dandelionR</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/dandelionR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_reproduce_original.html">Reproducing the original dandelion method/paper</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-cell immune repertoire trajectory
analysis with dandelionR</h1>
            
            <h4 data-toc-skip class="date">2025-02-14</h4>
      

      <div class="hidden name"><code>dandelionR.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>Welcome to <code>dandelionR</code>!</p>
<p><code>dandelionR</code> is an R package for performing single-cell
immune repertoire trajectory analysis, based on the original python
implementation in <a href="https://www.github.com/zktuong/dandelion" class="external-link">dandelion</a>.</p>
<p>It provides all the necessary tools to interface with <a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a>
and a custom implementation of absorbing markov chain for pseudotime
inference, inspired based on the <a href="https://github.com/dpeerlab/Palantir" class="external-link">palantir</a> python
package.</p>
<p>This is a work in progress, so please feel free to open an issue if
you encounter any problems or have any suggestions for improvement.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install <code>dandelionR</code> with:</p>
<div class="section level3">
<h3 id="bioconductor">Bioconductor<a class="anchor" aria-label="anchor" href="#bioconductor"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"dandelionR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="development-version-github">Development version (GitHub)<a class="anchor" aria-label="anchor" href="#development-version-github"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"tuonglab/dandelionR"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>In a standard analysis workflow in R, users probably choose to read
in their VDJ data with <a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a>.</p>
<p>In this vignette, we will demonstrate how to perform TCR trajectory
analysis starting from ‘raw’ data i.e. just a standard single-cell gene
expression data (stored in <code>SingleCellExperiment</code>) and VDJ
data (in AIRR format).</p>
<p>Install <code>scater</code> and <code>scRepertoire</code> if you
haven’t already.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># only for the tutorial</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"scater"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"scRepertoire"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scRepertoire"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># or</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ncborcherding/scRepertoire"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="load-the-required-libraries">Load the required libraries<a class="anchor" aria-label="anchor" href="#load-the-required-libraries"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.github.com/tuonglab/dandelionR/" class="external-link">dandelionR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.borch.dev/uploads/scRepertoire/" class="external-link">scRepertoire</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="load-the-demo-data">Load the demo data<a class="anchor" aria-label="anchor" href="#load-the-demo-data"></a>
</h3>
<p>Due to size limitations of the package, we have provided a very
trimmed down version of the demo data to ~2000 cells. The full dataset
can be found here accordingly:</p>
<p>GEX (Lymphoid Cells) - <a href="https://developmental.cellatlas.io/fetal-immune" class="external-link uri">https://developmental.cellatlas.io/fetal-immune</a></p>
<p>VDJ - <a href="https://github.com/zktuong/dandelion-demo-files/" class="external-link uri">https://github.com/zktuong/dandelion-demo-files/</a></p>
<p>The VDJ data is in the dandelion_manuscript/data/dandelion-remap
folder.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demo_sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">demo_airr</span><span class="op">)</span></span></code></pre></div>
<p>Check out the other vignette for an example dataset that starts from
the original <code>dandelion</code> output associated with the original
<a href="https://www.nature.com/articles/s41587-023-01734-7" class="external-link">manuscript</a>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"vignette_reproduce_original"</span><span class="op">)</span></span></code></pre></div>
<p>We will also set the seed so that the plots and results are
consistent.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="use-screpertoire-to-load-the-vdj-data">Use <code>scRepertoire</code> to load the VDJ data<a class="anchor" aria-label="anchor" href="#use-screpertoire-to-load-the-vdj-data"></a>
</h2>
<p>For the trajectory analysis work here, we are focusing on the main
productive TCR chains. Therefore we will flag
<code>filterMulti = TRUE</code>, which will keep the selection of the 2
corresponding chains with the highest expression for a single barcode.
For more details, refer to <code>scRepertoire</code>’s <a href="https://www.borch.dev/uploads/screpertoire/reference/combinetcr" class="external-link">docs</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">contig.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://www.borch.dev/uploads/scRepertoire/reference/loadContigs.html" class="external-link">loadContigs</a></span><span class="op">(</span>input <span class="op">=</span> <span class="va">demo_airr</span>, format <span class="op">=</span> <span class="st">"AIRR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Format to `scRepertoire`'s requirements and some light filtering</span></span>
<span><span class="va">combined.TCR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://www.borch.dev/uploads/scRepertoire/reference/combineTCR.html" class="external-link">combineTCR</a></span><span class="op">(</span><span class="va">contig.list</span>,</span>
<span>    removeNA <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    removeMulti <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    filterMulti <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="merging-vdj-data-with-gene-expression-data">Merging VDJ data with gene expression data<a class="anchor" aria-label="anchor" href="#merging-vdj-data-with-gene-expression-data"></a>
</h3>
<p>Next we will combine the gene expression data with the VDJ data to
create a <code>SingleCellExperiment</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://www.borch.dev/uploads/scRepertoire/reference/combineExpression.html" class="external-link">combineExpression</a></span><span class="op">(</span><span class="va">combined.TCR</span>, <span class="va">demo_sce</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="initiate-dandelionr-workflow">Initiate <code>dandelionR</code> workflow<a class="anchor" aria-label="anchor" href="#initiate-dandelionr-workflow"></a>
</h2>
<p>Here, the data is ready to be used for the pseudobulk and trajectory
analysis workflow in <code>dandelionR</code>.</p>
<p>Because this is a alpha-beta TCR data, we will set the
<code>mode_option</code> to “abT”. This will append <code>abT</code> to
the relevant columns holding the VDJ gene information. If you are going
to try other types of VDJ data e.g. BCR, you should set
<code>mode_option</code> to “B” instead. And this argument should be
consistently set with the <code>vdjPseudobulk</code> function later.</p>
<p>Since the TCR data is already filtered for productive chains in
<code>combineTCR</code>, we will set
<code>already.productive = TRUE</code> and can keep
<code>allowed_chain_status</code> as <code>NULL</code>.</p>
<p>We will also subset the data to only include the main T-cell types:
CD8+T, CD4+T, ABT(ENTRY), DP(P)_T, DP(Q)_T.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setupVdjPseudobulk.html">setupVdjPseudobulk</a></span><span class="op">(</span><span class="va">sce</span>,</span>
<span>    mode_option <span class="op">=</span> <span class="st">"abT"</span>,</span>
<span>    already.productive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    subsetby <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>,</span>
<span>    groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8+T"</span>, <span class="st">"CD4+T"</span>, <span class="st">"ABT(ENTRY)"</span>, <span class="st">"DP(P)_T"</span>, <span class="st">"DP(Q)_T"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The main output of this function is a
<code>SingleCellExperiment</code> object with the relevant VDJ
information appended to the <code>colData</code>, particularly the
columns with the <code>_main</code> suffix
e.g. <code>v_call_abT_VJ_main</code>, <code>j_call_abT_VJ_main</code>
etc.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 35 columns</span></span>
<span><span class="co">##                                  n_counts   n_genes           file      mito</span></span>
<span><span class="co">##                                 &lt;numeric&gt; &lt;integer&gt;       &lt;factor&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG      2947      1275 FCAImmP7851891 0.0105192</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG      4969      1971 FCAImmP7851892 0.0245522</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG      7230      1733 FCAImmP7803035 0.0302905</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA      2504       901 FCAImmP7528296 0.0207668</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA      8689      2037 FCAImmP7555860 0.0357924</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG      3111      1254 FCAImmP7292034 0.0228222</span></span>
<span><span class="co">##                                 doublet_scores predicted_doublets</span></span>
<span><span class="co">##                                      &lt;numeric&gt;           &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG      0.0439224              False</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG      0.0610687              False</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG      0.0383747              False</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA      0.0236220              False</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA      0.0738255              False</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG      0.0222841              False</span></span>
<span><span class="co">##                                 old_annotation_uniform    organ  Sort_id</span></span>
<span><span class="co">##                                               &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG              SP T CELL       TH    TOT  </span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG              DP T CELL       TH    TOT  </span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG              SP T CELL       SK    CD45P</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA              SP T CELL       SK    CD45P</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA              SP T CELL       TH    CD45P</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG              SP T CELL       TH    TOT  </span></span>
<span><span class="co">##                                       age   method    donor      sex</span></span>
<span><span class="co">##                                 &lt;integer&gt; &lt;factor&gt; &lt;factor&gt; &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG        11     5GEX      F64   female</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG        12     5GEX      F67   female</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG        14     5GEX      F51   female</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA        12     5GEX      F38   male  </span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA        16     5GEX      F41   female</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG        14     5GEX      F30   male  </span></span>
<span><span class="co">##                                                      Sample scvi_clusters</span></span>
<span><span class="co">##                                                    &lt;factor&gt;      &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG F64_TH_TOT_FCAImmP7851891              14</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG F67_TH_TOT_FCAImmP7851892              4 </span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG F51_SK_CD45P_FCAImmP7803035            2 </span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA F38_SK_CD45P_FCAImmP7528296            2 </span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA F41_TH_CD45P_FCAImmP7555860            2 </span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG F30_TH_TOT_FCAImmP7292034              14</span></span>
<span><span class="co">##                                 is_maternal_contaminant anno_lvl_2_final_clean</span></span>
<span><span class="co">##                                               &lt;logical&gt;               &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG                   FALSE             ABT(ENTRY)</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG                   FALSE             DP(Q)_T   </span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG                   FALSE             CD4+T     </span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA                   FALSE             CD4+T     </span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA                   FALSE             CD4+T     </span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG                   FALSE             ABT(ENTRY)</span></span>
<span><span class="co">##                                 celltype_annotation                 CTgene</span></span>
<span><span class="co">##                                            &lt;factor&gt;            &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG          ABT(ENTRY) TRAV13-1*02.TRAJ34*0..</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG          DP(Q)_T    TRAV12-2*01.TRAJ53*0..</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG          CD4+T      TRAV21*02.TRAJ33*01...</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA          CD4+T      TRAV38-2/DV8*01.TRAJ..</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA          CD4+T      TRAV12-1*01.TRAJ6*01..</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG          ABT(ENTRY) TRAV1-1*01.TRAJ32*02..</span></span>
<span><span class="co">##                                                   CTnt                   CTaa</span></span>
<span><span class="co">##                                            &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG TGTGCAGCAAGTATGAACAC.. CAASMNTDKLIF_CASSLTG..</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG TGTGCCGTGTGGAGGTAGCA.. CAVWR*QL*TDI_CASRTGN..</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG TGTGCTTCTATGGATAGCAA.. CASMDSNYQLIW_CASSLTS..</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA TGTGCTTATAGGAGCGTTCA.. CAYRSVQGAQKLVF_CASSW..</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA TGTGTGGTGAACATAAGAGG.. CVVNIRGSYIPTF_CSARDL..</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG TGCGCTGTGAGAGATCAGTA.. CAVRDQYGGATNKLIF_CAS..</span></span>
<span><span class="co">##                                               CTstrict clonalProportion</span></span>
<span><span class="co">##                                            &lt;character&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG TRAV13-1*02.TRAJ34*0..        0.0416667</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG TRAV12-2*01.TRAJ53*0..        0.0277778</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG TRAV21*02.TRAJ33*01...        0.0370370</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA TRAV38-2/DV8*01.TRAJ..        0.0909091</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA TRAV12-1*01.TRAJ6*01..        0.0476190</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG TRAV1-1*01.TRAJ32*02..        0.0204082</span></span>
<span><span class="co">##                                 clonalFrequency               cloneSize</span></span>
<span><span class="co">##                                       &lt;integer&gt;                &lt;factor&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG               1 Large (0.01 &lt; X &lt;= 0.1)</span></span>
<span><span class="co">##                                  v_call_abT_VDJ d_call_abT_VDJ j_call_abT_VDJ</span></span>
<span><span class="co">##                                     &lt;character&gt;    &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG     TRAV13-1*02      TRAJ34*01    TRBV11-3*04</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG     TRAV12-2*01      TRAJ53*01      TRBV19*01</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG       TRAV21*02      TRAJ33*01     TRBV5-4*01</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA TRAV38-2/DV8*01      TRAJ54*01     TRBV6-6*01</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA     TRAV12-1*01       TRAJ6*01    TRBV20-1*01</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG      TRAV1-1*01      TRAJ32*02    TRBV12-4*01</span></span>
<span><span class="co">##                                 v_call_abT_VJ j_call_abT_VJ v_call_abT_VDJ_main</span></span>
<span><span class="co">##                                   &lt;character&gt;   &lt;character&gt;         &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG      TRBD1*01    TRBJ1-2*01         TRAV13-1*02</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG            NA    TRBJ1-1*01         TRAV12-2*01</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG      TRBD2*02    TRBJ2-1*01           TRAV21*02</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA            NA    TRBJ1-3*01     TRAV38-2/DV8*01</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA      TRBD2*02    TRBJ2-7*01         TRAV12-1*01</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG            NA    TRBJ1-2*01          TRAV1-1*01</span></span>
<span><span class="co">##                                 d_call_abT_VDJ_main j_call_abT_VDJ_main</span></span>
<span><span class="co">##                                         &lt;character&gt;         &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG           TRAJ34*01         TRBV11-3*04</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG           TRAJ53*01           TRBV19*01</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG           TRAJ33*01          TRBV5-4*01</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA           TRAJ54*01          TRBV6-6*01</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA            TRAJ6*01         TRBV20-1*01</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG           TRAJ32*02         TRBV12-4*01</span></span>
<span><span class="co">##                                 v_call_abT_VJ_main j_call_abT_VJ_main</span></span>
<span><span class="co">##                                        &lt;character&gt;        &lt;character&gt;</span></span>
<span><span class="co">## FCAImmP7851891-CCTACCATCGGACAAG           TRBD1*01         TRBJ1-2*01</span></span>
<span><span class="co">## FCAImmP7851892-ACGGGCTCAGCATGAG                 NA         TRBJ1-1*01</span></span>
<span><span class="co">## FCAImmP7803035-CCAGCGATCCGAAGAG           TRBD2*02         TRBJ2-1*01</span></span>
<span><span class="co">## FCAImmP7528296-ATAAGAGTCAAAGACA                 NA         TRBJ1-3*01</span></span>
<span><span class="co">## FCAImmP7555860-AACTTTCTCAACGGGA           TRBD2*02         TRBJ2-7*01</span></span>
<span><span class="co">## FCAImmP7292034-CGTCACTGTGGTCTCG                 NA         TRBJ1-2*01</span></span></code></pre>
<p>Visualise the UMAP of the filtered data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="section level3">
<h3 id="milo-object-and-neighbourhood-graph-construction">Milo object and neighbourhood graph construction<a class="anchor" aria-label="anchor" href="#milo-object-and-neighbourhood-graph-construction"></a>
</h3>
<p>We will use miloR to create the pseudobulks based on the gene
expression data. The goal is to construct a neighbourhood graph with
many neighbors with which we can sample the representative neighbours to
form the objects.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marionilab.github.io/miloR" class="external-link">miloR</a></span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/Milo.html" class="external-link">Milo</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/buildGraph.html" class="external-link">buildGraph</a></span><span class="op">(</span><span class="va">milo_object</span>, k <span class="op">=</span> <span class="fl">30</span>, d <span class="op">=</span> <span class="fl">20</span>, reduced.dim <span class="op">=</span> <span class="st">"X_scvi"</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/makeNhoods.html" class="external-link">makeNhoods</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    reduced_dims <span class="op">=</span> <span class="st">"X_scvi"</span>, d <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    prop <span class="op">=</span> <span class="fl">0.3</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="construct-umap-on-milo-neighbor-graph">Construct UMAP on milo neighbor graph<a class="anchor" aria-label="anchor" href="#construct-umap-on-milo-neighbor-graph"></a>
</h3>
<p>We can visualise this milo object using UMAP.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/miloUmap.html">miloUmap</a></span><span class="op">(</span><span class="va">milo_object</span>, n_neighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>,</span>
<span>    dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="construct-pseudobulked-vdj-feature-space">Construct pseudobulked VDJ feature space<a class="anchor" aria-label="anchor" href="#construct-pseudobulked-vdj-feature-space"></a>
</h2>
<p>Next, we will construct the pseudobulked VDJ feature space using the
neighbourhood graph constructed above. We will also run PCA on the
pseudobulked VDJ feature space.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vdjPseudobulk.html">vdjPseudobulk</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    mode_option <span class="op">=</span> <span class="st">"abT"</span>,</span>
<span>    col_to_take <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Inspect the newly created <code>pb.milo</code> object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span></span></code></pre></div>
<pre><code><span><span class="co">## class: Milo </span></span>
<span><span class="co">## dim: 129 64 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): Feature_space</span></span>
<span><span class="co">## rownames(129): TRAV1-1*01 TRAV1-2*01 ... TRBJ2-7*01 TRBJ2-7*02</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(64): 179 93 ... 269 125</span></span>
<span><span class="co">## colData names(3): anno_lvl_2_final_clean</span></span>
<span><span class="co">##   anno_lvl_2_final_clean_fraction cell_count</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## nhoods dimensions(2): 64 382</span></span>
<span><span class="co">## nhoodCounts dimensions(2): 1 1</span></span>
<span><span class="co">## nhoodDistances dimension(1): 0</span></span>
<span><span class="co">## graph names(0):</span></span>
<span><span class="co">## nhoodIndex names(1): 0</span></span>
<span><span class="co">## nhoodExpression dimension(2): 1 1</span></span>
<span><span class="co">## nhoodReducedDim names(0):</span></span>
<span><span class="co">## nhoodGraph names(0):</span></span>
<span><span class="co">## nhoodAdjacency dimension(2): 1 1</span></span></code></pre>
<p>We can compute and visualise the PCA of the pseudobulked VDJ feature
space.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, assay.type <span class="op">=</span> <span class="st">"Feature_space"</span>, ncomponents <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="section level3">
<h3 id="tcr-trajectory-inference-using-absorbing-markov-chain">TCR trajectory inference using Absorbing Markov Chain<a class="anchor" aria-label="anchor" href="#tcr-trajectory-inference-using-absorbing-markov-chain"></a>
</h3>
<p>In the original <code>dandelion</code> python package, the trajectory
inference is done using the <code>palantir</code> package. Here, we
implement the absorbing markov chain approach in dandelionR to infer the
trajectory, leveraging on <code>destiny</code> for diffusion map
computation.</p>
<div class="section level4">
<h4 id="define-root-and-branch-tips">Define root and branch tips<a class="anchor" aria-label="anchor" href="#define-root-and-branch-tips"></a>
</h4>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the PCA matrix</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pb.milo</span>, type <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># define the CD8 terminal cell as the top-most cell and CD4 terminal cell as</span></span>
<span><span class="co"># the bottom-most cell</span></span>
<span><span class="va">branch.tips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">branch.tips</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8+T"</span>, <span class="st">"CD4+T"</span><span class="op">)</span></span>
<span><span class="co"># define the start of our trajectory as the right-most cell</span></span>
<span><span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="construct-diffusion-map">Construct diffusion map<a class="anchor" aria-label="anchor" href="#construct-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://theislab.github.io/destiny/" class="external-link">destiny</a></span><span class="op">)</span></span>
<span><span class="co"># Run diffusion map on the PCA</span></span>
<span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DiffusionMap-class.html" class="external-link">DiffusionMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span>, n_pcs <span class="op">=</span> <span class="fl">10</span>, n_eigs <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-diffusion-pseudotime-on-diffusion-map">Compute diffusion pseudotime on diffusion map<a class="anchor" aria-label="anchor" href="#compute-diffusion-pseudotime-on-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dif.pse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DPT.html" class="external-link">DPT</a></span><span class="op">(</span><span class="va">dm</span>, tips <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">root</span>, <span class="va">branch.tips</span><span class="op">)</span>, w_width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the root is automatically called DPT + index of the root cell</span></span>
<span><span class="va">DPTroot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"DPT"</span>, <span class="va">root</span><span class="op">)</span></span>
<span><span class="co"># store pseudotime in milo object</span></span>
<span><span class="va">pb.milo</span><span class="op">$</span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="va">dif.pse</span><span class="op">[[</span><span class="va">DPTroot</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># set the colours for pseudotime</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="markov-chain-construction-on-the-pseudobulk-vdj-feature-space">Markov chain construction on the pseudobulk VDJ feature space<a class="anchor" aria-label="anchor" href="#markov-chain-construction-on-the-pseudobulk-vdj-feature-space"></a>
</h2>
<p>This step will compute the Markov chain probabilities on the
pseudobulk VDJ feature space. It will return the branch probabilities in
the <code>colData</code> and the column name corresponds to the branch
tips defined earlier.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markovProbability.html">markovProbability</a></span><span class="op">(</span></span>
<span>    milo <span class="op">=</span> <span class="va">pb.milo</span>,</span>
<span>    diffusionmap <span class="op">=</span> <span class="va">dm</span>,</span>
<span>    terminal_state <span class="op">=</span> <span class="va">branch.tips</span>,</span>
<span>    root_cell <span class="op">=</span> <span class="va">root</span>,</span>
<span>    pseudotime_key <span class="op">=</span> <span class="st">"pseudotime"</span>,</span>
<span>    knn <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Inspect the <code>pb.milo</code> object to see the newly added
columns.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 6 columns</span></span>
<span><span class="co">##     anno_lvl_2_final_clean anno_lvl_2_final_clean_fraction cell_count</span></span>
<span><span class="co">##                &lt;character&gt;                       &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## 179              faDP(P)_T                        0.605263         38</span></span>
<span><span class="co">## 93                 faCD8+T                        0.750000         48</span></span>
<span><span class="co">## 75               faDP(Q)_T                        0.907407         54</span></span>
<span><span class="co">## 279              faDP(Q)_T                        0.913793         58</span></span>
<span><span class="co">## 277                faCD4+T                        0.936170         47</span></span>
<span><span class="co">## 375              faDP(P)_T                        0.632653         49</span></span>
<span><span class="co">##     pseudotime     CD8+T     CD4+T</span></span>
<span><span class="co">##      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 179   0.449589  0.578761  0.421239</span></span>
<span><span class="co">## 93    5.584408  0.729279  0.270721</span></span>
<span><span class="co">## 75    2.642788  0.559456  0.440544</span></span>
<span><span class="co">## 279   2.595964  0.559617  0.440383</span></span>
<span><span class="co">## 277   4.314940  0.565749  0.434251</span></span>
<span><span class="co">## 375   0.134566  0.578744  0.421256</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualising-branch-probabilities">Visualising branch probabilities<a class="anchor" aria-label="anchor" href="#visualising-branch-probabilities"></a>
</h2>
<p>With the Markov chain probabilities computed, we can visualise the
branch probabilities towards CD4+ or CD8+ T-cell fate on the PCA
plot.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"CD8+T"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"CD4+T"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-26-2.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="transfer">Transfer<a class="anchor" aria-label="anchor" href="#transfer"></a>
</h2>
<p>The next step is to project the pseudotime and the branch probability
information from the pseudobulks back to each cell in the dataset. If
the cell do not belong to any of the pseudobulk, it will be removed. If
a cell belongs to multiple pseudobulk samples, its value should be
calculated as a weighted average of the corresponding values from each
pseudobulk, where each weight is inverse of the size of the
pseudobulk.</p>
<div class="section level3">
<h3 id="project-pseudobulk-data-to-each-cell">Project pseudobulk data to each cell<a class="anchor" aria-label="anchor" href="#project-pseudobulk-data-to-each-cell"></a>
</h3>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/projectPseudotimeToCell.html">projectPseudotimeToCell</a></span><span class="op">(</span><span class="va">milo_object</span>, <span class="va">pb.milo</span>, <span class="va">branch.tips</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualise-the-trajectory-data-on-a-per-cell-basis">Visualise the trajectory data on a per cell basis<a class="anchor" aria-label="anchor" href="#visualise-the-trajectory-data-on-a-per-cell-basis"></a>
</h3>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-28-2.png" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD4+T"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-28-3.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD8+T"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-28-4.png" width="700"></p>
<p>And that’s it! We have successfully inferred the trajectory of the
T-cells in this dataset!</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2025-02-12 r87715)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] destiny_3.21.0              miloR_2.3.0                </span></span>
<span><span class="co">##  [3] edgeR_4.5.2                 limma_3.63.3               </span></span>
<span><span class="co">##  [5] scater_1.35.1               scuttle_1.17.0             </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.29.1 SummarizedExperiment_1.37.0</span></span>
<span><span class="co">##  [9] Biobase_2.67.0              GenomicRanges_1.59.1       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.43.4         IRanges_2.41.3             </span></span>
<span><span class="co">## [13] S4Vectors_0.45.4            BiocGenerics_0.53.6        </span></span>
<span><span class="co">## [15] generics_0.1.3              MatrixGenerics_1.19.1      </span></span>
<span><span class="co">## [17] matrixStats_1.5.0           scRepertoire_2.3.2         </span></span>
<span><span class="co">## [19] ggplot2_3.5.1               dandelionR_0.99.10         </span></span>
<span><span class="co">## [21] BiocStyle_2.35.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] cubature_2.1.1          splines_4.5.0           tibble_3.2.1           </span></span>
<span><span class="co">##   [4] polyclip_1.10-7         xts_0.14.1              lifecycle_1.0.4        </span></span>
<span><span class="co">##   [7] globals_0.16.3          lattice_0.22-6          MASS_7.3-64            </span></span>
<span><span class="co">##  [10] magrittr_2.0.3          vcd_1.4-13              sass_0.4.9             </span></span>
<span><span class="co">##  [13] rmarkdown_2.29          jquerylib_0.1.4         yaml_2.3.10            </span></span>
<span><span class="co">##  [16] spam_2.11-1             sp_2.2-0                cowplot_1.1.3          </span></span>
<span><span class="co">##  [19] RColorBrewer_1.1-3      abind_1.4-8             purrr_1.0.4            </span></span>
<span><span class="co">##  [22] ggraph_2.2.1            nnet_7.3-20             pracma_2.4.4           </span></span>
<span><span class="co">##  [25] tweenr_2.0.3            evmix_2.12              GenomeInfoDbData_1.2.13</span></span>
<span><span class="co">##  [28] ggrepel_0.9.6           irlba_2.3.5.1           listenv_0.9.1          </span></span>
<span><span class="co">##  [31] iNEXT_3.0.1             MatrixModels_0.5-3      RSpectra_0.16-2        </span></span>
<span><span class="co">##  [34] parallelly_1.42.0       pkgdown_2.1.1           codetools_0.2-20       </span></span>
<span><span class="co">##  [37] smoother_1.3            DelayedArray_0.33.5     ggforce_0.4.2          </span></span>
<span><span class="co">##  [40] tidyselect_1.2.1        UCSC.utils_1.3.1        farver_2.1.2           </span></span>
<span><span class="co">##  [43] ScaledMatrix_1.15.0     viridis_0.6.5           jsonlite_1.8.9         </span></span>
<span><span class="co">##  [46] BiocNeighbors_2.1.2     e1071_1.7-16            tidygraph_1.3.1        </span></span>
<span><span class="co">##  [49] progressr_0.15.1        Formula_1.2-5           survival_3.8-3         </span></span>
<span><span class="co">##  [52] ggalluvial_0.12.5       systemfonts_1.2.1       tools_4.5.0            </span></span>
<span><span class="co">##  [55] ragg_1.3.3              stringdist_0.9.15       Rcpp_1.0.14            </span></span>
<span><span class="co">##  [58] glue_1.8.0              gridExtra_2.3           SparseArray_1.7.5      </span></span>
<span><span class="co">##  [61] laeken_0.5.3            xfun_0.50               ranger_0.17.0          </span></span>
<span><span class="co">##  [64] TTR_0.24.4              ggthemes_5.1.0          dplyr_1.1.4            </span></span>
<span><span class="co">##  [67] withr_3.0.2             numDeriv_2016.8-1.1     BiocManager_1.30.25    </span></span>
<span><span class="co">##  [70] fastmap_1.2.0           boot_1.3-31             bluster_1.17.0         </span></span>
<span><span class="co">##  [73] SparseM_1.84-2          VIM_6.2.2               digest_0.6.37          </span></span>
<span><span class="co">##  [76] rsvd_1.0.5              R6_2.6.0                textshaping_1.0.0      </span></span>
<span><span class="co">##  [79] colorspace_2.1-1        gtools_3.9.5            tidyr_1.3.1            </span></span>
<span><span class="co">##  [82] hexbin_1.28.5           data.table_1.16.4       robustbase_0.99-4-1    </span></span>
<span><span class="co">##  [85] class_7.3-23            graphlayouts_1.2.2      httr_1.4.7             </span></span>
<span><span class="co">##  [88] htmlwidgets_1.6.4       S4Arrays_1.7.3          scatterplot3d_0.3-44   </span></span>
<span><span class="co">##  [91] uwot_0.2.2              pkgconfig_2.0.3         gtable_0.3.6           </span></span>
<span><span class="co">##  [94] lmtest_0.9-40           XVector_0.47.2          htmltools_0.5.8.1      </span></span>
<span><span class="co">##  [97] carData_3.0-5           dotCall64_1.2           bookdown_0.42          </span></span>
<span><span class="co">## [100] SeuratObject_5.0.2      scales_1.3.0            knn.covertree_1.0      </span></span>
<span><span class="co">## [103] ggdendro_0.2.0          knitr_1.49              rjson_0.2.23           </span></span>
<span><span class="co">## [106] reshape2_1.4.4          curl_6.2.0              proxy_0.4-27           </span></span>
<span><span class="co">## [109] cachem_1.1.0            zoo_1.8-12              stringr_1.5.1          </span></span>
<span><span class="co">## [112] parallel_4.5.0          vipor_0.4.7             desc_1.4.3             </span></span>
<span><span class="co">## [115] pillar_1.10.1           grid_4.5.0              vctrs_0.6.5            </span></span>
<span><span class="co">## [118] pcaMethods_1.99.0       VGAM_1.1-13             car_3.1-3              </span></span>
<span><span class="co">## [121] BiocSingular_1.23.0     beachmat_2.23.6         cluster_2.1.8          </span></span>
<span><span class="co">## [124] beeswarm_0.4.0          evaluate_1.0.3          truncdist_1.0-2        </span></span>
<span><span class="co">## [127] cli_3.6.4               locfit_1.5-9.11         compiler_4.5.0         </span></span>
<span><span class="co">## [130] rlang_1.1.5             crayon_1.5.3            future.apply_1.11.3    </span></span>
<span><span class="co">## [133] labeling_0.4.3          plyr_1.8.9              fs_1.6.5               </span></span>
<span><span class="co">## [136] ggbeeswarm_0.7.2        stringi_1.8.4           viridisLite_0.4.2      </span></span>
<span><span class="co">## [139] BiocParallel_1.41.0     assertthat_0.2.1        munsell_0.5.1          </span></span>
<span><span class="co">## [142] gsl_2.1-8               quantreg_6.00           Matrix_1.7-2           </span></span>
<span><span class="co">## [145] RcppHNSW_0.6.0          RcppEigen_0.3.4.0.2     patchwork_1.3.0        </span></span>
<span><span class="co">## [148] future_1.34.0           statmod_1.5.0           evd_2.3-7.1            </span></span>
<span><span class="co">## [151] igraph_2.1.4            memoise_2.0.1           bslib_0.9.0            </span></span>
<span><span class="co">## [154] DEoptimR_1.1-3-1        ggplot.multistats_1.0.1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jiawei Yu, Nicholas Borcherding, Kelvin Tuong.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
