<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single-cell Immune Repertoire Trajectory Analysis with dandelionR â€¢ dandelionR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell Immune Repertoire Trajectory Analysis with dandelionR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dandelionR</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Unreleased version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/dandelionR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Single-cell Immune Repertoire Trajectory Analysis with dandelionR</h1>
            
            <h4 data-toc-skip class="date">2024-12-09</h4>
      

      <div class="d-none name"><code>dandelionR.rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>Welcome to <code>dandelionR</code>!</p>
<p><code>dandelionR</code> is an R package for performing single-cell
immune repertoire trajectory analysis, based on the original python
implementation in <a href="https://www.github.com/zktuong/dandelion" class="external-link">dandelion</a>.</p>
<p>It provides all the necessary tools to interface with <a href="https://github.com/ncborcherding/scRepertoire" class="external-link">scRepertoire</a>
and a custom implementation of absorbing markov chain for pseudotime
inference, inspired based on the <a href="https://github.com/dpeerlab/Palantir" class="external-link">palantir</a> python
package.</p>
<div class="section level3">
<h3 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h3>
<p>You can install <code>dandelionR</code> from GitHub with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"devtools"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"scater"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="co"># only for the tutorial</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scater"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"tuonglab/dandelionR"</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tuonglab.github.io/dandelionR/" class="external-link">dandelionR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
<p>This is a work in progress, so please feel free to open an issue if
you encounter any problems or have any suggestions for improvement.</p>
</div>
<div class="section level3">
<h3 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h3>
<p>First, we will load the demo data. This is a down-sampled dataset
from the <a href="https://www.nature.com/articles/s41587-023-01734-7" class="external-link">Suo et al 2024
Nature Biotechnology</a> paper. It contains 32000 cells with the TCR
information and the dimensionality reduced data (scVI) that we need for
this tutorial. The gene expression matrix is not required for this
tutorial so it is not included in the demo data. We will show in a
separate tutorial how to start with data from
<code>scRepertoire</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sce_vdj</span><span class="op">)</span></span></code></pre></div>
<p>We will set the seed so that the plots and results are
consistent.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filter-the-data">Filter the data<a class="anchor" aria-label="anchor" href="#filter-the-data"></a>
</h3>
<p>To begin, we will filter the data and extract the TCR information so
that we can construct pseudobulks.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setupVdjPseudobulk.html">setupVdjPseudobulk</a></span><span class="op">(</span><span class="va">sce_vdj</span>, already.productive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We can visualise the UMAP of the filtered data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_vdj</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="milo-object-and-neighbourhood-graph-construction">Milo object and neighbourhood graph construction<a class="anchor" aria-label="anchor" href="#milo-object-and-neighbourhood-graph-construction"></a>
</h3>
<p>We will use miloR to create the pseudobulks based on the gene
expression data. The goal is to construct a neighbourhood graph with
many neighbors with which we can sample the representative neighbours to
form the objects.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marionilab.github.io/miloR" class="external-link">miloR</a></span><span class="op">)</span></span>
<span><span class="va">traj_milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/Milo.html" class="external-link">Milo</a></span><span class="op">(</span><span class="va">sce_vdj</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/buildGraph.html" class="external-link">buildGraph</a></span><span class="op">(</span><span class="va">traj_milo</span>, k <span class="op">=</span> <span class="fl">50</span>, d <span class="op">=</span> <span class="fl">20</span>, reduced.dim <span class="op">=</span> <span class="st">"X_scvi"</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/makeNhoods.html" class="external-link">makeNhoods</a></span><span class="op">(</span><span class="va">milo_object</span>, reduced_dims <span class="op">=</span> <span class="st">"X_scvi"</span>, d <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="construct-umap-on-milo-neighbor-graph">Construct UMAP on milo neighbor graph<a class="anchor" aria-label="anchor" href="#construct-umap-on-milo-neighbor-graph"></a>
</h4>
<p>We can visualise this milo object using UMAP.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/miloUmap.html">miloUmap</a></span><span class="op">(</span><span class="va">milo_object</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">milo_object</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="construct-pseudobulked-vdj-feature-space">Construct pseudobulked VDJ feature space<a class="anchor" aria-label="anchor" href="#construct-pseudobulked-vdj-feature-space"></a>
</h3>
<p>Next, we will construct the pseudobulked VDJ feature space using the
neighbourhood graph constructed above. We will also run PCA on the
pseudobulked VDJ feature space.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vdjPseudobulk.html">vdjPseudobulk</a></span><span class="op">(</span><span class="va">milo_object</span>, col_to_take <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># pbs = milo_object@nhoods</span></span>
<span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, assay.type <span class="op">=</span> <span class="st">"Feature_space"</span><span class="op">)</span></span></code></pre></div>
<p>We can visualise the PCA of the pseudobulked VDJ feature space.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="tcr-trajectory-inference-using-absorbing-markov-chain">TCR trajectory inference using Absorbing Markov Chain<a class="anchor" aria-label="anchor" href="#tcr-trajectory-inference-using-absorbing-markov-chain"></a>
</h3>
<p>In the original <code>dandelion</code> python package, the trajectory
inference is done using the <code>palantir</code> package. Here, we
implement the absorbing markov chain approach in dandelionR to infer the
trajectory, leveraging on <code>destiny</code> for diffusion map
computation.</p>
<div class="section level4">
<h4 id="define-root-and-branch-tips">Define root and branch tips<a class="anchor" aria-label="anchor" href="#define-root-and-branch-tips"></a>
</h4>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract the PCA matrix</span></span>
<span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">pb.milo</span>, type <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># define the CD8 terminal cell as the top-most cell and CD4 terminal cell as the bottom-most cell</span></span>
<span><span class="va">branch.tips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">branch.tips</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8+T"</span>, <span class="st">"CD4+T"</span><span class="op">)</span></span>
<span><span class="co"># define the start of our trajectory as the left-most cell</span></span>
<span><span class="va">root</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="construct-diffusion-map">Construct diffusion map<a class="anchor" aria-label="anchor" href="#construct-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://theislab.github.io/destiny/" class="external-link">destiny</a></span><span class="op">)</span></span>
<span><span class="co"># Run diffusion map on the PCA</span></span>
<span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DiffusionMap-class.html" class="external-link">DiffusionMap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pca</span><span class="op">)</span>, n_pcs <span class="op">=</span> <span class="fl">50</span>, n_eigs <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="compute-diffussion-pseudotime-on-diffusion-map">Compute diffussion pseudotime on diffusion map<a class="anchor" aria-label="anchor" href="#compute-diffussion-pseudotime-on-diffusion-map"></a>
</h4>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dif.pse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/destiny/man/DPT.html" class="external-link">DPT</a></span><span class="op">(</span><span class="va">dm</span>, tips <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">root</span>, <span class="va">branch.tips</span><span class="op">)</span>, w_width <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the root is automatically called DPT + index of the root cell</span></span>
<span><span class="va">DPTroot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"DPT"</span>, <span class="va">root</span><span class="op">)</span></span>
<span><span class="co"># store pseudotime in milo object</span></span>
<span><span class="va">pb.milo</span><span class="op">$</span><span class="va">pseudotime</span> <span class="op">&lt;-</span> <span class="va">dif.pse</span><span class="op">[[</span><span class="va">DPTroot</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># set the colours for pseudotime</span></span>
<span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level4">
<h4 id="markov-chain-construction-on-the-pseudobulk-vdj-feature-space">Markov chain construction on the pseudobulk VDJ feature space<a class="anchor" aria-label="anchor" href="#markov-chain-construction-on-the-pseudobulk-vdj-feature-space"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markovProbability.html">markovProbability</a></span><span class="op">(</span></span>
<span>    milo <span class="op">=</span> <span class="va">pb.milo</span>,</span>
<span>    diffusionmap <span class="op">=</span> <span class="va">dm</span>,</span>
<span>    terminal_state <span class="op">=</span> <span class="va">branch.tips</span>,</span>
<span>    root_cell <span class="op">=</span> <span class="va">root</span>,</span>
<span>    pseudotime_key <span class="op">=</span> <span class="st">"pseudotime"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualising-branch-probabilities">Visualising branch probabilities<a class="anchor" aria-label="anchor" href="#visualising-branch-probabilities"></a>
</h4>
<p>With the Markov chain probabilities computed, we can visualise the
branch probabilities towards CD4+ or CD8+ T-cell fate on the PCA
plot.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"CD8+T"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"CD4+T"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-18-2.png" width="700"></p>
</div>
</div>
<div class="section level3">
<h3 id="transfer">Transfer<a class="anchor" aria-label="anchor" href="#transfer"></a>
</h3>
<p>The next step is to project the pseudotime and the branch probability
information from the pseudobulks back to each cell in the dataset. If
the cell do not belong to any of the pseudobulk, it will be removed. If
a cell belongs to multiple pseudobulk samples, its value should be
calculated as a weighted average of the corresponding values from each
pseudobulk, where each weight is inverse of the size of the
pseudobulk.</p>
<div class="section level4">
<h4 id="project-pseudobulk-data-to-each-cell">Project pseudobulk data to each cell<a class="anchor" aria-label="anchor" href="#project-pseudobulk-data-to-each-cell"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/projectPseudotimeToCell.html">projectPseudotimeToCell</a></span><span class="op">(</span><span class="va">milo_object</span>, <span class="va">pb.milo</span>, <span class="va">branch.tips</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="visualise-the-trajectory-data-on-a-per-cell-basis">Visualise the trajectory data on a per cell basis<a class="anchor" aria-label="anchor" href="#visualise-the-trajectory-data-on-a-per-cell-basis"></a>
</h4>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"pseudotime"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD4+T"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-20-3.png" width="700"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"CD8+T"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_files/figure-html/unnamed-chunk-20-4.png" width="700"></p>
<p>And thatâ€™s it! We have successfully inferred the trajectory of the
T-cells in this dataset!</p>
</div>
</div>
<div class="section level3">
<h3 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] destiny_3.20.0              miloR_2.2.0                </span></span>
<span><span class="co">##  [3] edgeR_4.4.1                 limma_3.62.1               </span></span>
<span><span class="co">##  [5] scater_1.34.0               ggplot2_3.5.1              </span></span>
<span><span class="co">##  [7] scuttle_1.16.0              SingleCellExperiment_1.28.1</span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.36.0 Biobase_2.66.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.58.0        GenomeInfoDb_1.42.1        </span></span>
<span><span class="co">## [13] IRanges_2.40.0              S4Vectors_0.44.0           </span></span>
<span><span class="co">## [15] BiocGenerics_0.52.0         MatrixGenerics_1.18.0      </span></span>
<span><span class="co">## [17] matrixStats_1.4.1           dandelionR_0.99.0          </span></span>
<span><span class="co">## [19] BiocStyle_2.34.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.5                spatstat.sparse_3.1-0   httr_1.4.7             </span></span>
<span><span class="co">##   [4] RColorBrewer_1.1-3      numDeriv_2016.8-1.1     tools_4.4.2            </span></span>
<span><span class="co">##   [7] sctransform_0.4.1       utf8_1.2.4              R6_2.5.1               </span></span>
<span><span class="co">##  [10] lazyeval_0.2.2          uwot_0.2.2              withr_3.0.2            </span></span>
<span><span class="co">##  [13] sp_2.1-4                gridExtra_2.3           progressr_0.15.1       </span></span>
<span><span class="co">##  [16] cli_3.6.3               textshaping_0.4.1       spatstat.explore_3.3-3 </span></span>
<span><span class="co">##  [19] fastDummies_1.7.4       labeling_0.4.3          sass_0.4.9             </span></span>
<span><span class="co">##  [22] Seurat_5.1.0            robustbase_0.99-4-1     spatstat.data_3.1-4    </span></span>
<span><span class="co">##  [25] proxy_0.4-27            ggridges_0.5.6          pbapply_1.7-2          </span></span>
<span><span class="co">##  [28] pkgdown_2.1.1           systemfonts_1.1.0       parallelly_1.40.1      </span></span>
<span><span class="co">##  [31] TTR_0.24.4              generics_0.1.3          gtools_3.9.5           </span></span>
<span><span class="co">##  [34] ica_1.0-3               spatstat.random_3.3-2   car_3.1-3              </span></span>
<span><span class="co">##  [37] dplyr_1.1.4             Matrix_1.7-1            ggbeeswarm_0.7.2       </span></span>
<span><span class="co">##  [40] fansi_1.0.6             abind_1.4-8             lifecycle_1.0.4        </span></span>
<span><span class="co">##  [43] scatterplot3d_0.3-44    yaml_2.3.10             carData_3.0-5          </span></span>
<span><span class="co">##  [46] SparseArray_1.6.0       Rtsne_0.17              grid_4.4.2             </span></span>
<span><span class="co">##  [49] promises_1.3.2          crayon_1.5.3            miniUI_0.1.1.1         </span></span>
<span><span class="co">##  [52] lattice_0.22-6          beachmat_2.22.0         cowplot_1.1.3          </span></span>
<span><span class="co">##  [55] pillar_1.9.0            knitr_1.49              boot_1.3-31            </span></span>
<span><span class="co">##  [58] future.apply_1.11.3     codetools_0.2-20        leiden_0.4.3.1         </span></span>
<span><span class="co">##  [61] glue_1.8.0              spatstat.univar_3.1-1   pcaMethods_1.98.0      </span></span>
<span><span class="co">##  [64] data.table_1.16.4       vcd_1.4-13              vctrs_0.6.5            </span></span>
<span><span class="co">##  [67] png_0.1-8               spam_2.11-0             gtable_0.3.6           </span></span>
<span><span class="co">##  [70] cachem_1.1.0            xfun_0.49               S4Arrays_1.6.0         </span></span>
<span><span class="co">##  [73] mime_0.12               RcppEigen_0.3.4.0.2     tidygraph_1.3.1        </span></span>
<span><span class="co">##  [76] pracma_2.4.4            survival_3.7-0          bluster_1.16.0         </span></span>
<span><span class="co">##  [79] statmod_1.5.0           fitdistrplus_1.2-1      ROCR_1.0-11            </span></span>
<span><span class="co">##  [82] nlme_3.1-166            xts_0.14.1              RcppAnnoy_0.0.22       </span></span>
<span><span class="co">##  [85] bslib_0.8.0             irlba_2.3.5.1           vipor_0.4.7            </span></span>
<span><span class="co">##  [88] KernSmooth_2.23-24      colorspace_2.1-1        nnet_7.3-19            </span></span>
<span><span class="co">##  [91] smoother_1.3            tidyselect_1.2.1        curl_6.0.1             </span></span>
<span><span class="co">##  [94] compiler_4.4.2          BiocNeighbors_2.0.1     desc_1.4.3             </span></span>
<span><span class="co">##  [97] DelayedArray_0.32.0     plotly_4.10.4           scales_1.3.0           </span></span>
<span><span class="co">## [100] hexbin_1.28.5           DEoptimR_1.1-3-1        lmtest_0.9-40          </span></span>
<span><span class="co">## [103] stringr_1.5.1           digest_0.6.37           goftest_1.2-3          </span></span>
<span><span class="co">## [106] spatstat.utils_3.1-1    rmarkdown_2.29          XVector_0.46.0         </span></span>
<span><span class="co">## [109] htmltools_0.5.8.1       pkgconfig_2.0.3         fastmap_1.2.0          </span></span>
<span><span class="co">## [112] ggthemes_5.1.0          rlang_1.1.4             htmlwidgets_1.6.4      </span></span>
<span><span class="co">## [115] UCSC.utils_1.2.0        shiny_1.9.1             farver_2.1.2           </span></span>
<span><span class="co">## [118] jquerylib_0.1.4         zoo_1.8-12              jsonlite_1.8.9         </span></span>
<span><span class="co">## [121] BiocParallel_1.40.0     BiocSingular_1.22.0     magrittr_2.0.3         </span></span>
<span><span class="co">## [124] Formula_1.2-5           GenomeInfoDbData_1.2.13 dotCall64_1.2          </span></span>
<span><span class="co">## [127] patchwork_1.3.0         munsell_0.5.1           Rcpp_1.0.13-1          </span></span>
<span><span class="co">## [130] viridis_0.6.5           reticulate_1.40.0       stringi_1.8.4          </span></span>
<span><span class="co">## [133] ggraph_2.2.1            zlibbioc_1.52.0         MASS_7.3-61            </span></span>
<span><span class="co">## [136] plyr_1.8.9              parallel_4.4.2          listenv_0.9.1          </span></span>
<span><span class="co">## [139] ggrepel_0.9.6           deldir_2.0-4            graphlayouts_1.2.1     </span></span>
<span><span class="co">## [142] splines_4.4.2           tensor_1.5              locfit_1.5-9.10        </span></span>
<span><span class="co">## [145] igraph_2.1.1            ranger_0.17.0           spatstat.geom_3.3-4    </span></span>
<span><span class="co">## [148] RcppHNSW_0.6.0          reshape2_1.4.4          ScaledMatrix_1.14.0    </span></span>
<span><span class="co">## [151] evaluate_1.0.1          SeuratObject_5.0.2      BiocManager_1.30.25    </span></span>
<span><span class="co">## [154] laeken_0.5.3            tweenr_2.0.3            httpuv_1.6.15          </span></span>
<span><span class="co">## [157] VIM_6.2.2               RANN_2.6.2              tidyr_1.3.1            </span></span>
<span><span class="co">## [160] purrr_1.0.2             polyclip_1.10-7         knn.covertree_1.0      </span></span>
<span><span class="co">## [163] future_1.34.0           scattermore_1.2         ggforce_0.4.2          </span></span>
<span><span class="co">## [166] rsvd_1.0.5              xtable_1.8-4            e1071_1.7-16           </span></span>
<span><span class="co">## [169] RSpectra_0.16-2         later_1.4.1             viridisLite_0.4.2      </span></span>
<span><span class="co">## [172] class_7.3-22            ragg_1.3.3              tibble_3.2.1           </span></span>
<span><span class="co">## [175] memoise_2.0.1           beeswarm_0.4.0          cluster_2.1.6          </span></span>
<span><span class="co">## [178] ggplot.multistats_1.0.1 globals_0.16.3</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jiawei Yu, Nicholas Borcherding, Kelvin Tuong.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
