<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Single-cell immune repertoire trajectory analysis with dandelionR and slingshot • dandelionR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell immune repertoire trajectory analysis with dandelionR and slingshot">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">dandelionR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/dandelionR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/dandelionR_with_slingshot.html">Single-cell immune repertoire trajectory analysis with dandelionR and slingshot</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_reproduce_original.html">Reproducing the original dandelion method/paper</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Single-cell immune repertoire trajectory analysis with dandelionR and slingshot</h1>
            
            <h4 data-toc-skip class="date">2025-08-08</h4>
      

      <div class="d-none name"><code>dandelionR_with_slingshot.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="foreword">Foreword<a class="anchor" aria-label="anchor" href="#foreword"></a>
</h2>
<p>In this vignette, we will demonstrate how to perform TCR trajectory
analysis using the <code>dandelionR</code> package in conjunction with
<code>slingshot</code>.</p>
</div>
<div class="section level2">
<h2 id="installing-slinshot">Installing slinshot<a class="anchor" aria-label="anchor" href="#installing-slinshot"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"slinghot"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"slingshot"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="section level3">
<h3 id="load-the-required-libraries-and-data">Load the required libraries and data<a class="anchor" aria-label="anchor" href="#load-the-required-libraries-and-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.github.com/tuonglab/dandelionR/" class="external-link">dandelionR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.borch.dev/uploads/scRepertoire/" class="external-link">scRepertoire</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">sce_vdj</span><span class="op">)</span></span></code></pre></div>
<p>We will also set the seed so that the plots and results are
consistent.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setup-the-data-as-per-the-other-vignettes">Setup the data as per the other vignettes<a class="anchor" aria-label="anchor" href="#setup-the-data-as-per-the-other-vignettes"></a>
</h3>
<p>We will use the full <code>SingleCellExperiment</code> object
<code>sce_vdj</code> that contains enough cells to demonstrate the
trajectory analysis with <code>slingshot</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce_vdj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/setupVdjPseudobulk.html">setupVdjPseudobulk</a></span><span class="op">(</span><span class="va">sce_vdj</span>,</span>
<span>    already.productive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    allowed_chain_status <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"Single pair"</span>, <span class="st">"Extra pair"</span>,</span>
<span>        <span class="st">"Extra pair-exception"</span>, <span class="st">"Orphan VDJ"</span>,</span>
<span>        <span class="st">"Orphan VDJ-exception"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce_vdj</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="milo-object-and-neighbourhood-graph-construction">Milo object and neighbourhood graph construction<a class="anchor" aria-label="anchor" href="#milo-object-and-neighbourhood-graph-construction"></a>
</h3>
<p>We will use miloR to create the pseudobulks based on the gene
expression data. The goal is to construct a neighbourhood graph with
many neighbors with which we can sample the representative neighbours to
form the objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marionilab.github.io/miloR" class="external-link">miloR</a></span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/Milo.html" class="external-link">Milo</a></span><span class="op">(</span><span class="va">sce_vdj</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/buildGraph.html" class="external-link">buildGraph</a></span><span class="op">(</span><span class="va">milo_object</span>, k <span class="op">=</span> <span class="fl">30</span>, d <span class="op">=</span> <span class="fl">20</span>, reduced.dim <span class="op">=</span> <span class="st">"X_scvi"</span><span class="op">)</span></span>
<span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://marionilab.github.io/miloR/reference/makeNhoods.html" class="external-link">makeNhoods</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    reduced_dims <span class="op">=</span> <span class="st">"X_scvi"</span>, d <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    prop <span class="op">=</span> <span class="fl">0.3</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="construct-umap-on-milo-neighbor-graph">Construct UMAP on milo neighbor graph<a class="anchor" aria-label="anchor" href="#construct-umap-on-milo-neighbor-graph"></a>
</h3>
<p>We can visualise this milo object using UMAP.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">milo_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/miloUmap.html">miloUmap</a></span><span class="op">(</span><span class="va">milo_object</span>, n_neighbors <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>,</span>
<span>    dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="construct-pseudobulked-vdj-feature-space">Construct pseudobulked VDJ feature space<a class="anchor" aria-label="anchor" href="#construct-pseudobulked-vdj-feature-space"></a>
</h2>
<p>Next, we will construct the pseudobulked VDJ feature space using the
neighbourhood graph constructed above. We will also run PCA on the
pseudobulked VDJ feature space.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/vdjPseudobulk.html">vdjPseudobulk</a></span><span class="op">(</span><span class="va">milo_object</span>,</span>
<span>    mode_option <span class="op">=</span> <span class="st">"abT"</span>,</span>
<span>    col_to_take <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Inspect the newly created <code>pb.milo</code> object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span></span></code></pre></div>
<pre><code><span><span class="co">## class: Milo </span></span>
<span><span class="co">## dim: 151 499 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): Feature_space</span></span>
<span><span class="co">## rownames(151): TRBV10-1 TRBV10-2 ... TRAJ8 TRAJ9</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(499): 2463 147 ... 977 533</span></span>
<span><span class="co">## colData names(3): anno_lvl_2_final_clean</span></span>
<span><span class="co">##   anno_lvl_2_final_clean_fraction cell_count</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## nhoods dimensions(2): 499 2657</span></span>
<span><span class="co">## nhoodCounts dimensions(2): 1 1</span></span>
<span><span class="co">## nhoodDistances dimension(1): 0</span></span>
<span><span class="co">## graph names(0):</span></span>
<span><span class="co">## nhoodIndex names(1): 0</span></span>
<span><span class="co">## nhoodExpression dimension(2): 1 1</span></span>
<span><span class="co">## nhoodReducedDim names(0):</span></span>
<span><span class="co">## nhoodGraph names(0):</span></span>
<span><span class="co">## nhoodAdjacency dimension(2): 1 1</span></span></code></pre>
<p>We can compute and visualise the PCA of the pseudobulked VDJ feature
space.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, assay.type <span class="op">=</span> <span class="st">"Feature_space"</span>, ncomponents <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">pb.milo</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="section level3">
<h3 id="tcr-trajectory-inference-using-slingshot">TCR trajectory inference using Slingshot<a class="anchor" aria-label="anchor" href="#tcr-trajectory-inference-using-slingshot"></a>
</h3>
<p>In the original dandelion Python package, trajectory inference is
performed using the palantir package. Here, we instead use the Slingshot
package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">slingshot</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="input">input<a class="anchor" aria-label="anchor" href="#input"></a>
</h4>
<p>Slingshot requires a matrix representing cells in a
reduced-dimensional space and a vector of cluster labels. Here, we use
PCA for dimensionality reduction and the column anno_lvl_2_final_clean
from colData as the cluster labels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span><span class="op">[[</span><span class="st">"anno_lvl_2_final_clean"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## faABT(ENTRY)      faCD4+T      faCD8+T    faDP(P)_T    faDP(Q)_T </span></span>
<span><span class="co">##           98          185          106           49           61</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="run-slingshot">run slingshot<a class="anchor" aria-label="anchor" href="#run-slingshot"></a>
</h4>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pb.milo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/slingshot.html" class="external-link">slingshot</a></span><span class="op">(</span><span class="va">pb.milo</span>, clusterLabels <span class="op">=</span> <span class="fu">colData</span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span><span class="op">[[</span><span class="st">"anno_lvl_2_final_clean"</span><span class="op">]</span><span class="op">]</span>, reducedDim <span class="op">=</span> <span class="st">"PCA"</span>, start.clus <span class="op">=</span> <span class="st">"faDP(P)_T"</span>, end.clus <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"faCD4+T"</span>, <span class="st">"faCD8+T"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>pseudotime of each lineage, if the pseudobulk is with value NA, it
will not appear on the phage.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dnychka/fieldsRPackage" class="external-link">fields</a></span><span class="op">)</span></span>
<span><span class="co"># make a color palette from blue to red</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#d7191c"</span>, <span class="st">"#fdae61"</span>, <span class="st">"#ffffbf"</span>, <span class="st">"#abd9e9"</span>, <span class="st">"#2c7bb6"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span><span class="op">$</span><span class="va">PCA</span>, col <span class="op">=</span> <span class="va">colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">$</span><span class="va">slingPseudotime_1</span>, breaks <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, asp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/SlingshotDataSet.html" class="external-link">SlingshotDataSet</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fields/man/image.plot.html" class="external-link">image.plot</a></span><span class="op">(</span></span>
<span>    legend.only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">$</span><span class="va">slingPseudotime_1</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>    legend.lab <span class="op">=</span> <span class="st">"Pseudotime"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDims</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span><span class="op">$</span><span class="va">PCA</span>, col <span class="op">=</span> <span class="va">colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">$</span><span class="va">slingPseudotime_2</span>, breaks <span class="op">=</span> <span class="fl">50</span><span class="op">)</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, asp <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/slingshot/man/SlingshotDataSet.html" class="external-link">SlingshotDataSet</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/fields/man/image.plot.html" class="external-link">image.plot</a></span><span class="op">(</span></span>
<span>    legend.only <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    zlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="va">pb.milo</span><span class="op">$</span><span class="va">slingPseudotime_2</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>    legend.lab <span class="op">=</span> <span class="st">"Pseudotime"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="transfer">Transfer<a class="anchor" aria-label="anchor" href="#transfer"></a>
</h2>
<p>The next step is to project the pseudotime information from the
pseudobulks back to each cell in the dataset. If the cell do not belong
to any of the pseudobulk, it will be removed. If a cell belongs to
multiple pseudobulk samples, its value should be calculated as a
weighted average of the corresponding values from each pseudobulk, where
each weight is inverse of the size of the pseudobulk. Pseudobulks with
NA values are excluded during the projection. If all pseudobulks
associated with a cell have NA values, the projected value for that cell
will also be NA. ## Project pseudobulk data to each cell</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/projectPseudotimeToCell.html">projectPseudotimeToCell</a></span><span class="op">(</span><span class="va">milo_object</span>, <span class="va">pb.milo</span>, value_key <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"slingPseudotime_1"</span>, <span class="st">"slingPseudotime_2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="visualise-the-trajectory-data-on-a-per-cell-basis">Visualise the trajectory data on a per cell basis<a class="anchor" aria-label="anchor" href="#visualise-the-trajectory-data-on-a-per-cell-basis"></a>
</h3>
<p>The cell with NA value will in grey.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">(</span><span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"RdYlBu"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"anno_lvl_2_final_clean"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"slingPseudotime_1"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-18-2.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">cdata</span>, color_by <span class="op">=</span> <span class="st">"slingPseudotime_2"</span>, dimred <span class="op">=</span> <span class="st">"UMAP_knngraph"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="va">pal</span><span class="op">)</span></span></code></pre></div>
<p><img src="dandelionR_with_slingshot_files/figure-html/unnamed-chunk-18-3.png" width="700"></p>
<p>And that’s it! We have successfully inferred the trajectory of the
T-cells in this dataset with <code>slingshot</code>!</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] fields_16.3.1               viridisLite_0.4.2          </span></span>
<span><span class="co">##  [3] spam_2.11-1                 slingshot_2.16.0           </span></span>
<span><span class="co">##  [5] TrajectoryUtils_1.16.1      princurve_2.1.6            </span></span>
<span><span class="co">##  [7] miloR_2.4.1                 edgeR_4.6.3                </span></span>
<span><span class="co">##  [9] limma_3.64.3                scater_1.36.0              </span></span>
<span><span class="co">## [11] scuttle_1.18.0              SingleCellExperiment_1.30.1</span></span>
<span><span class="co">## [13] SummarizedExperiment_1.38.1 Biobase_2.68.0             </span></span>
<span><span class="co">## [15] GenomicRanges_1.60.0        GenomeInfoDb_1.44.1        </span></span>
<span><span class="co">## [17] IRanges_2.42.0              S4Vectors_0.46.0           </span></span>
<span><span class="co">## [19] BiocGenerics_0.54.0         generics_0.1.4             </span></span>
<span><span class="co">## [21] MatrixGenerics_1.20.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [23] scRepertoire_2.4.0          ggplot2_3.5.2              </span></span>
<span><span class="co">## [25] dandelionR_1.1.4            BiocStyle_2.36.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] cubature_2.1.4            splines_4.5.1            </span></span>
<span><span class="co">##   [3] tibble_3.3.0              polyclip_1.10-7          </span></span>
<span><span class="co">##   [5] xts_0.14.1                lifecycle_1.0.4          </span></span>
<span><span class="co">##   [7] globals_0.18.0            lattice_0.22-7           </span></span>
<span><span class="co">##   [9] MASS_7.3-65               magrittr_2.0.3           </span></span>
<span><span class="co">##  [11] vcd_1.4-13                sass_0.4.10              </span></span>
<span><span class="co">##  [13] rmarkdown_2.29            jquerylib_0.1.4          </span></span>
<span><span class="co">##  [15] yaml_2.3.10               sp_2.2-0                 </span></span>
<span><span class="co">##  [17] cowplot_1.2.0             RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [19] maps_3.4.3                abind_1.4-8              </span></span>
<span><span class="co">##  [21] purrr_1.1.0               ggraph_2.2.1             </span></span>
<span><span class="co">##  [23] nnet_7.3-20               pracma_2.4.4             </span></span>
<span><span class="co">##  [25] tweenr_2.0.3              evmix_2.12               </span></span>
<span><span class="co">##  [27] GenomeInfoDbData_1.2.14   ggrepel_0.9.6            </span></span>
<span><span class="co">##  [29] irlba_2.3.5.1             listenv_0.9.1            </span></span>
<span><span class="co">##  [31] iNEXT_3.0.2               MatrixModels_0.5-4       </span></span>
<span><span class="co">##  [33] RSpectra_0.16-2           parallelly_1.45.1        </span></span>
<span><span class="co">##  [35] DelayedMatrixStats_1.30.0 pkgdown_2.1.3            </span></span>
<span><span class="co">##  [37] codetools_0.2-20          smoother_1.3             </span></span>
<span><span class="co">##  [39] DelayedArray_0.34.1       ggforce_0.5.0            </span></span>
<span><span class="co">##  [41] tidyselect_1.2.1          UCSC.utils_1.4.0         </span></span>
<span><span class="co">##  [43] farver_2.1.2              ScaledMatrix_1.16.0      </span></span>
<span><span class="co">##  [45] viridis_0.6.5             jsonlite_2.0.0           </span></span>
<span><span class="co">##  [47] BiocNeighbors_2.2.0       e1071_1.7-16             </span></span>
<span><span class="co">##  [49] tidygraph_1.3.1           progressr_0.15.1         </span></span>
<span><span class="co">##  [51] Formula_1.2-5             survival_3.8-3           </span></span>
<span><span class="co">##  [53] ggalluvial_0.12.5         systemfonts_1.2.3        </span></span>
<span><span class="co">##  [55] tools_4.5.1               ragg_1.4.0               </span></span>
<span><span class="co">##  [57] stringdist_0.9.15         Rcpp_1.1.0               </span></span>
<span><span class="co">##  [59] glue_1.8.0                gridExtra_2.3            </span></span>
<span><span class="co">##  [61] SparseArray_1.8.1         laeken_0.5.3             </span></span>
<span><span class="co">##  [63] xfun_0.52                 ranger_0.17.0            </span></span>
<span><span class="co">##  [65] TTR_0.24.4                ggthemes_5.1.0           </span></span>
<span><span class="co">##  [67] dplyr_1.1.4               withr_3.0.2              </span></span>
<span><span class="co">##  [69] numDeriv_2016.8-1.1       BiocManager_1.30.26      </span></span>
<span><span class="co">##  [71] fastmap_1.2.0             boot_1.3-31              </span></span>
<span><span class="co">##  [73] bluster_1.18.0            SparseM_1.84-2           </span></span>
<span><span class="co">##  [75] VIM_6.2.2                 digest_0.6.37            </span></span>
<span><span class="co">##  [77] rsvd_1.0.5                R6_2.6.1                 </span></span>
<span><span class="co">##  [79] textshaping_1.0.1         colorspace_2.1-1         </span></span>
<span><span class="co">##  [81] gtools_3.9.5              tidyr_1.3.1              </span></span>
<span><span class="co">##  [83] hexbin_1.28.5             data.table_1.17.8        </span></span>
<span><span class="co">##  [85] robustbase_0.99-4-1       class_7.3-23             </span></span>
<span><span class="co">##  [87] graphlayouts_1.2.2        httr_1.4.7               </span></span>
<span><span class="co">##  [89] htmlwidgets_1.6.4         S4Arrays_1.8.1           </span></span>
<span><span class="co">##  [91] scatterplot3d_0.3-44      uwot_0.2.3               </span></span>
<span><span class="co">##  [93] pkgconfig_2.0.3           gtable_0.3.6             </span></span>
<span><span class="co">##  [95] lmtest_0.9-40             XVector_0.48.0           </span></span>
<span><span class="co">##  [97] destiny_3.22.0            htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [99] carData_3.0-5             dotCall64_1.2            </span></span>
<span><span class="co">## [101] bookdown_0.43             SeuratObject_5.1.0       </span></span>
<span><span class="co">## [103] scales_1.4.0              ggdendro_0.2.0           </span></span>
<span><span class="co">## [105] knitr_1.50                rjson_0.2.23             </span></span>
<span><span class="co">## [107] reshape2_1.4.4            curl_6.4.0               </span></span>
<span><span class="co">## [109] proxy_0.4-27              cachem_1.1.0             </span></span>
<span><span class="co">## [111] zoo_1.8-14                stringr_1.5.1            </span></span>
<span><span class="co">## [113] parallel_4.5.1            vipor_0.4.7              </span></span>
<span><span class="co">## [115] desc_1.4.3                pillar_1.11.0            </span></span>
<span><span class="co">## [117] grid_4.5.1                vctrs_0.6.5              </span></span>
<span><span class="co">## [119] RANN_2.6.2                pcaMethods_2.0.0         </span></span>
<span><span class="co">## [121] VGAM_1.1-13               car_3.1-3                </span></span>
<span><span class="co">## [123] BiocSingular_1.24.0       beachmat_2.24.0          </span></span>
<span><span class="co">## [125] cluster_2.1.8.1           beeswarm_0.4.0           </span></span>
<span><span class="co">## [127] evaluate_1.0.4            truncdist_1.0-2          </span></span>
<span><span class="co">## [129] cli_3.6.5                 locfit_1.5-9.12          </span></span>
<span><span class="co">## [131] compiler_4.5.1            rlang_1.1.6              </span></span>
<span><span class="co">## [133] crayon_1.5.3              future.apply_1.20.0      </span></span>
<span><span class="co">## [135] labeling_0.4.3            plyr_1.8.9               </span></span>
<span><span class="co">## [137] fs_1.6.6                  ggbeeswarm_0.7.2         </span></span>
<span><span class="co">## [139] stringi_1.8.7             BiocParallel_1.42.1      </span></span>
<span><span class="co">## [141] assertthat_0.2.1          gsl_2.1-8                </span></span>
<span><span class="co">## [143] quantreg_6.1              Matrix_1.7-3             </span></span>
<span><span class="co">## [145] RcppHNSW_0.6.0            RcppEigen_0.3.4.0.2      </span></span>
<span><span class="co">## [147] patchwork_1.3.1           sparseMatrixStats_1.20.0 </span></span>
<span><span class="co">## [149] future_1.67.0             statmod_1.5.0            </span></span>
<span><span class="co">## [151] evd_2.3-7.1               igraph_2.1.4             </span></span>
<span><span class="co">## [153] memoise_2.0.1             bslib_0.9.0              </span></span>
<span><span class="co">## [155] DEoptimR_1.1-4            ggplot.multistats_1.0.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jiawei Yu, Nicholas Borcherding, Kelvin Tuong.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
